### PR TITLE
feat: Show avatar images

### PR DESCRIPTION
**Pull request by nesium**
This PR introduces the ability to upload an avatar image and display avatars of participants in a chat and people in your roster.

I've introduced two state containers `SessionState` and `ChatSessionState`. These carry state about the current user and the selected chat (See rationale [here](https://forums.swift.org/t/best-practice-for-sharing-data-between-many-features/37696/4)) and can be shared between reducers.

It's now possible to go into the "Edit profile" screen and upload a new avatar image by dragging & dropping it onto the avatar field in the top left.

I'm sorry that this grew so large but I had to take several shortcuts already:

- There's no activity indicator nor error handling when uploading the avatar. Once you see the new image however you know that everything went well.
- Avatar metadata and images are loaded eagerly as soon as the ProseCoreClient receives the roster. Images are cached between restarts however.
- Images are uploaded with a maximum side-length of 600px.
- Downloaded images are only cached in one size with a maximum side-length of 300px.
- Resizing before the upload could be improved by distinguishing if the dragged item was a file URL or an image and then resizing the image via ImageIO and not loading the whole thing at once into memory.
- I've chosen to ignore reducer lifecycle warnings at this time. I didn't want to introduce yet another concept in this PR.
- Avatars from other users are not updated in real-time when changed. You need to restart to app to load the new ones. This is because I couldn't get to reliably receive pubsub notifications. marc@prose.org  -> cram@prose.org worked without a problem, but not the other way around. I have to investigate that separately.

### CODE REVIEW

**Comment by valeriansaliou** [comment]
Just tested the feature, works & looks great for me! 💯

**Comment by RemiBardon** [comment]
Thank you very much 🙏🏼
I will review this PR this morning.

**Review by RemiBardon — APPROVED** [review]
Wow, that was some reading! 📖

Thank you for all of this!

I added some edit suggestions, but overall things look great 🙂
I haven't tested the code, but Valerian did and the CI passes, so I believe it's fine.

File: Prose/ProseLib/Sources/Toolbox/CGImage+Prose.swift [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+public extension CGImage {
+  func prose_downsampled(maxPixelSize: Int) throws -> CGImage {
+    let imageSize = CGSize(
+      width: self.width,
+      height: self.height
+    )
+
+    guard Int(imageSize.width) > maxPixelSize || Int(imageSize.height) > maxPixelSize else {
```
[the remark this answers is not in the document]
Comment (nesium) [reply Prose/ProseLib/Sources/Toolbox/CGImage+Prose.swift:70]: That's nicer.

File: Prose/ProseLib/Sources/Toolbox/CGImage+Prose.swift [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+      newSize = CGSize(
+        width: CGFloat(maxPixelSize),
+        height: imageSize.height / imageSize.width * CGFloat(maxPixelSize)
+      )
+    default:
+      newSize = CGSize(
+        width: imageSize.width / imageSize.height * CGFloat(maxPixelSize),
+        height: CGFloat(maxPixelSize)
+      )
+    }
```
[the remark this answers is not in the document]
Comment (nesium) [reply Prose/ProseLib/Sources/Toolbox/CGImage+Prose.swift]: I've used this, just reformatted with 2 spaces.

File: Prose/ProseLib/Sources/ProseCoreTCA/AvatarImageCache.swift [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+public extension AvatarImageCache {
+  static func live(cacheDirectory: URL) throws -> Self {
+    let fileManager = FileManager.default
+
+    try fileManager.createDirectory(at: cacheDirectory, withIntermediateDirectories: true)
+
+    func fileURL(for jid: JID, imageId: String) -> URL {
+      let filename = "\(jid.bareJid.node ?? "")-\(jid.bareJid.domain)-\(imageId)".unicodeScalars
+        .filter(CharacterSet.urlPathAllowed.contains)
```
[the remark this answers is not in the document]
Comment (nesium) [reply Prose/ProseLib/Sources/ProseCoreTCA/AvatarImageCache.swift:29]: Good catch! I wish there was a method on `URL` though. Just give me a path that is safe to use with the filesystem. I don't know how often I've looked that up in the past.

File: Prose/ProseLib/Sources/Toolbox/CGImage+Prose.swift [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+    let destinationProperties = [
+      kCGImageDestinationLossyCompressionQuality: compressionQuality,
+    ] as CFDictionary
+
+    CGImageDestinationAddImage(imageDestination, self, destinationProperties)
+    CGImageDestinationFinalize(imageDestination)
+
+    return buffer as Data
+  }
```
[the remark this answers is not in the document]
Comment (nesium) [reply Prose/ProseLib/Sources/Toolbox/CGImage+Prose.swift:128]: In this case I'll pick the solution that works cross-platform, even if it's a few lines longer. That means that this particular code path is hit more often and has a greater chance to be correct, even after a change.

File: Prose/ProseLib/Sources/Toolbox/CGImage+Prose.swift [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+enum DownsamplingError: Error {
+  case couldNotDownsampleImage
+  case couldNotCreateImageDestination
+  case invalidInputImage
+  case couldNotFinalizeImage
+}
+
+public extension CGImageSource {
+  func prose_downsample(
```
[the remark this answers is not in the document]
Comment (nesium) [reply Prose/ProseLib/Sources/Toolbox/CGImage+Prose.swift:24]: After watching the [WWDC talk](https://developer.apple.com/videos/play/wwdc2018/219/) back, I'm conviced that using Core Graphics is the right approach. I've added a todo note in the NSItemProvider extension that we should resize the images in the future using Core Graphics before loading them into memory.

**Pull request merged**